_pages/dev/ssr/server-side-rendering-error-handling.md

```diff
@@ -0,0 +1,422 @@
+- make sure that `provideServer()` config function is provided in the `app.server.module.ts` file. It contains elements required for SSR Error Handling to work properly.
+
+**Note:** The SSR Error Handling is available starting from Spartacus 2211.29. For new application, the feature is enabled fy default. For migrated application. the feature is disabled by default and can be enabled by using mentioned feature toggles. After expiration of the feature toggle period, the feature will be enabled by default.
+**Note:** Together with SSR Error Handling, Spartacus provides `ssrFeatureToggles` that allow you to enable or disable features specific for `OptimizedSsrEngine`. For more information, see [SSR Feature Toggles](TODO: add link to page when is read).
+
+Without SSR Error Handling, the following issues arise:
+
+- with the default Angular SSR behavior (used under the hood of Spartacus SSR), when the asynchronous error occurs during the rendering (e.g. API call, runtime error in the code), rendering process will not stop and eventually returns possibly malformed HTML to the client with the improper HTTP status code `200`. This can lead to the client-side errors and the page not being rendered correctly.
+- when customer enters the URL unknown for the application, Spartacus returns an error page, but with misleading status `200`.
+
+Both scenarios may drastically affect the SEO of the page due to fact the wrong status code in the response to the client might lead to indexing by Google the malformed pages and unknown URLs.
+
+See below the sequence diagram of rendering the Angular app for an incoming request in SSR, with Spartacus OptimizedSsrEngine included. It shows how an asynchronous error happening during the rendering is ignored and a malformed HTML is returned to a client:
+![Rendering Sequence Without SSR Error Handling](../../../assets/images/ssr/error_handling_rendering_sequence_without_improvements.png)
+
+To solve the issues, Spartacus provided CxCommonEngine - a wrapper for CommonEngine which allows to react on asynchronous errors that occurred during SSR. Thanks to this, it is possible to return an error with the appropriate status code instead of the rendered HTML which is potentially malformed.
+
+See below the sequence diagrams showing how the path from the incoming request to the error response looks with the attached CxCommonEngine:
+
+![Rendering Sequence With SSR Error Handling](../../../assets/images/ssr/error_handling_rendering_sequence_with_improvements.png)
+
+To propagate errors from the Angular app to ExpressJS, Spartacus introduce a new contract between the Angular app and ExpressJS.
+See below the diagram showing the connections between elements of the contract:
+
+![SSR Error Handling Contract](../../../assets/images/ssr/error_handling_contract_between_angular_and_expressjs.png)
```

Comment:
The diagram contains an outdated name (before it got renamed in the source code):
old: `defaultServerErrorResponseHandlers`
new: `defaultExpressErrorHandlers`